storage/bolt/bolt.go

```diff
@@ -0,0 +1,236 @@
+package bolt
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/boltdb/bolt"
+
+	"github.com/lovoo/goka/storage"
+)
+
+const (
+	offsetKey = "__offset"
+)
+
+type boltStorage struct {
+	client     *bolt.DB
+	bucketName []byte
+}
+
+// New creates a new Storage backed by Redis.
+func New(client *bolt.DB, bucket string) (storage.Storage, error) {
+	if client == nil {
+		return nil, errors.New("invalid bolt db")
+	}
+
+	tx, err := client.Begin(true)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	// Use the transaction...
+	tx.CreateBucket([]byte(bucket))
+	if err != nil {
+		return nil, err
+	}
+
+	// Commit the transaction and check for error.
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+
+	return &boltStorage{
+		client:     client,
+		bucketName: []byte(bucket),
+	}, nil
+}
+
+func (b *boltStorage) Has(key string) (bool, error) {
+	hasKey := false
+	err := b.client.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(b.bucketName)
+		k := []byte(key)
+		value := b.Get(k)
+		if value != nil {
+			hasKey = true
+		}
+		return nil
+	})
+
+	return hasKey, err
+}
+
+func (b *boltStorage) Get(key string) ([]byte, error) {
+	var value []byte
+	err := b.client.View(func(tx *bolt.Tx) error {
+		var err error
+		buck := tx.Bucket(b.bucketName)
+		k := []byte(key)
+		value = buck.Get(k)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	return value, err
+}
+
+func (b *boltStorage) GetOffset(defValue int64) (int64, error) {
+	data, err := b.Get(offsetKey)
+	if err != nil {
+		return 0, err
+	}
+	if data == nil {
+		return defValue, nil
+	}
+
+	value, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding bolt offset (%s): %v", string(data), err)
+	}
+	return value, nil
+}
+
+func (b *boltStorage) Set(key string, value []byte) error {
+	err := b.client.Update(func(tx *bolt.Tx) error {
```

Comment:
Probably we need to distinguish between recovering mode and running mode, otherwise it'll create a new transaction for every new value in recovery, which probably slows it down. But let's optimize later.